### ISSUE TITLE
Save files are absurdly huge

### ISSUE THREAD

**Issue by ShadowNinja** [pioneerspacesim/pioneer#3680]
I made a few saves and noticed that the game locked up for a few seconds when saving/loading, I just ignored this at the time, but later I looked at my save files and noticed that they were on the order of 30MiB and growing with each save!

The entire file is apparently serialized as JSON, which partially explains the save delay (although even something faster like MsgPack is going to have trouble with 30 MiB of structured data).  The rest of the delay is probably just the time it takes to write that much data to disk (`dd` takes about 2.2s to write 30MiB).

After looking through the file it looks like a lot of large sections are duplicated dozens of times.  Additionally, it looks like some values stored as strings are really entire sub-files with many lines of their own.

The save file is too big for me to really get a good idea on what it's structure is without going over the code that generated it, so I've had to use bulk text management tools like `grep`.

I don't know exactly what Pioneer is saving, but given the simple actions I'm taking in the game it shouldn't be even 10 KiB, let alone 30+ MiB!

Version in question: 2016-04-04

**Comment by impaktor** [comment]
Not a bug, as this is the way things are since we moved to json format. If you want to improve upon it, then you're welcome. The person who wrote the json save code did not have detailed knowledge of the code.

**Comment by laarmen** [comment]
IIRC the previous save file weren't exactly slim either (a few MB, I
believe).

I acknowledge the problem of the size of our save files

However, your last remark (given the few actions, etc) shows a deep
misunderstanding of how the game works at its core. Pretty all content in
game is procedurally generated from an RNG seeded with a determined seed,
which means all players play in the same galaxy. However, pretty much any
player interaction can potentially consume some values off the RNG, which
means the state of the whole universe (well, most likely the current
system, but you get the idea) depends on those few actions.

From there, two options: we can record the whole game state at save time
and reload it plain and simple, or "just" record the disruptive actions.
(and that's all assuming the state depends uniquely on the one RNG, which
I'm not even sure of) and recompute the whole world. Even with the single
RNG assumption, this method means exploding save times the further you get
into the game. As you may have guessed, we've opted for (a version of) the
first solution. We can maybe get halfway on some parts (for instance, when
storing Lua objects, we could provide the (de)serializer with an
"archetypical" object holding the most common values, and store a diff of
it.

All-in-all, the less we store the more likely we are to load a world that
was not the one saved because of evolutions in the code. The more we store,
the more likely we are to break the savegame, period :-)

**Comment by ShadowNinja** [comment]
I make no claims to understanding this game's architecture, after all I've only been playing with it for a day.  :-)

Saving only object fields that have deviated from the default should help a bit.

I've started going over the code and so far I've noticed these things could be improved:
* If serialized `lua_Number`s are integers don't write out the `.000000` in LuaSerializer::pickle.
* Perhaps replace newlines with a different character (LuaSerializer).
* `"lua_modules"` is serialized with `BinStrToJson`, which serializes the string as an array of ints -- multiplying the size of the binary string about 8x.  Just serializing it as a string with escapes would be better, and the data is text data anyway, so this is completely unnecessary.
* `"trailing_signature"` can be removed.
* Use sequential ints (or even base 16/base 64 strings) for table IDs (as opposed to floats).

**Comment by laarmen** [comment]
Oh, well, you know what they say, "if you've got an itch scratch it";
"patches welcome", and all that jazz ! I actually think that's a nice way
to start contributing.

**Comment by ShadowNinja** [comment]
I'm making some general improvements, but it's hard to see the bigger picture and figure out what things should be removed/restructured when the file is so big that it just eats up a few GiB of memory and crashes in all the JSON tree viewers I've tried so far.

**Comment by impaktor** [comment]
@ShadowNinja Might be of interest to see the full path of how Json saves came to be: [dev.forum](http://pioneerspacesim.net/forum/viewtopic.php?f=3&t=263). 

Also, if you want an account on dev. forum, just say the word, and we'll fix it. Registration is currently closed due to T1000's roaming about.

**Comment by ShadowNinja** [comment]
I've noticed that floating point values are often stored by casting their memory value to an unsigned int and then writing that to a string.  Besides being nonportable, this is less efficient and the result is unreadable.  `StrToFloat` mentions that this is done so that no precision is lost, but it this really an issue?  The loss of precision will only be something like 1/1,000,000.  That may not be perfect, but It seems close enough.  Floating point math is barely deterministic as it is anyways.

I've also noticed that Frames (and SFX) get double-nested like so:
```JSON
// A Frame, currently:
{
	"child_frames": [
		{
			"frame": {
				"<Frame fields...>"
			}
		}, {
			"frame": {
				"<Frame fields...>"
			}
		}
	]
}
// How it could be:
{
	"child_frames": [
		{
			"<Frame fields...>"
		}, {
			"<Frame fields...>"
		}
	]
}
```

It may also be worth looking into a different serialization format for frames and bodies, as most of their size is just overhead (field names, punctuation, the float-uint-string thing, etc).

**Comment by robn** [comment]
> I've noticed that floating point values are often stored by casting their memory value to an unsigned int and then writing that to a string. Besides being nonportable, this is less efficient and the result is unreadable. StrToFloat mentions that this is done so that no precision is lost, but it this really an issue? The loss of precision will only be something like 1/1,000,000. That may not be perfect, but It seems close enough. Floating point math is barely deterministic as it is anyways.

* It doesn't have to be portable
* Preserving precision is _critical_

More generally, I'm not sure what your complaint is. Are you concerned about size, load/save time, or readability?

Size I don't really care about, but could be trivially improved with some kind of compression.

Load/save time is mostly about getting the internal state into a serialisable form, and would be about the same no matter what format used (and indeed, was about the same on the old binary savefile format). There are improvements that can be made but they're non-trivial in the extreme.

Readability I don't care about. The JSON format is there to be useful for developers to look at, and to assist with upgrading (because there's a little more flexibility in the format).

If you'd like to make improvements, they would be very welcome, but please do take the time to understand the reasons behind what already exists.

**Comment by ShadowNinja** [comment]
@robn:
Why is precision so critical?  And if it's so critical why isn't the same done for the LuaSerialization format?  I'm just trying to understand whether the original motivation for the code is still relevant.

I'm concerned about all three (size, performance, readability), but primarily the first two.

You say that the old binary format's performance was just as bad as the JSON format.  Did it generate files of similar size, or were they much smaller?

I do understand that JSON was added for backwards compatibility and flexibility.  I'm maintaining backwards compatibility with my changes and I'm actually making it more flexible (by making fields optional, one of the advantages of a key-value system).

Can you please keep this issue open?  It isn't fixed yet, and it seems like a legitimate issue, at least to me.

**Comment by fluffyfreak** [comment]
Reopening to track :)

**Comment by fluffyfreak** [comment]
Just to clarify the precision part. It's important for core because it's the position of bodies on planets within systems.
This isn't always so critical but there's many times, especially on larger world's, where 1/1,000,000th displacement can mean a few hundred metre/kilometres difference.

It's probably less of an issue with most lua modules. Or it has been so far anyway.

**Comment by ShadowNinja** [comment]
I've noticed that save files created in 2016-04-04 won't load in latest git, without any of my changes.  Here's the error:
```
pioneer: SpaceStation.cpp:201: void SpaceStation::InitStation(): Assertion `m_shipDocking.size() == m_type->NumDockingPorts()' failed.
```
It seems like the code depends on the RNG always returning the same values so that the randomly selected stations match up, and there's a mismatch.  Sometimes several of the stations match up before a mismatching one is hit.

Also, I've made progress making the file smaller:  It is now about 3 MiB, and compression brings it down to 300 KiB (with gzip/zlib deflate, LZMA gets down to 60 KiB).

The biggest space saver so far is probably replacing `BinStrToJson` (for `"lua_modules"`) with a simple string.
Other changes I've done:
* Switch from `Json::StyledWriter` to `Json::FastWriter` (I'd argue that the fancy indentation isn't really useful in a file this large anyways).
* Make some fields optional.
* Remove extra nesting previously mentioned.
* Remove `"trailing_signature"`, and `.000000`; and make table reference numbers sequential.

**Comment by fluffyfreak** [comment]
@ShadowNinja which system does that save game loading break in?

**Comment by fluffyfreak** [comment]
Nevermind I've found that I can get it with just the Earth starting point. Which is worrying.

**Comment by fluffyfreak** [comment]
@ShadowNinja I've bene looking into this as well, the `BinStrToJson` cannot be a raw string as it has both unprintable characters and can mess with the JSON. Sometimes it works by luck, other times it fails.

It'd be good to see this branch you're working on.

**Comment by impaktor** [comment]
@fluffyfreak  closed by #3684 ?

**Comment by fluffyfreak** [comment]
Ah yes.
